File: plugin_src/browsertoolbar/src/android/TitleBarIconView.java
package org.elastos.essentials.plugins.browsertoolbar;

import android.content.Context;
import android.graphics.Color;
import android.graphics.ColorMatrixColorFilter;
import android.net.Uri;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.ImageView;

import org.elastos.essentials.app.R;

public class TitleBarIconView extends FrameLayout {
    ImageButton ivMainIcon;
    ImageView ivBadge;

    public TitleBarIconView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        LayoutInflater inflater = LayoutInflater.from(getContext());
        inflater.inflate(R.layout.title_bar_icon_view, this, true);

        ivMainIcon = findViewById(R.id.ivMainIcon);
        ivBadge = findViewById(R.id.ivBadge);

        setBadgeCount(0);
    }

    public float convertDpToPx(Context context, float dp) {
        return dp * context.getResources().getDisplayMetrics().density;
    }

    public float convertPxToDp(Context context, float px) {
        return px / context.getResources().getDisplayMetrics().density;
    }

    public void setPaddingDp(int dpPadding) {
        int paddingPx = (int) convertDpToPx(getContext(), dpPadding);
        ivMainIcon.setPadding(paddingPx, paddingPx, paddingPx, paddingPx);
    }

    public void setImageResource(int resId) {
        ivMainIcon.setImageResource(resId);
    }

    public void setImageURI(Uri uri) {
        ivMainIcon.setImageURI(uri);
    }

    public void setColorFilter(int color) {
        ivMainIcon.setColorFilter(color);
    }

    public void setColorFilter(ColorMatrixColorFilter filter) {
        ivMainIcon.setColorFilter(filter);
    }

    public void setOnClickListener(OnClickListener listener) {
        ivMainIcon.setOnClickListener(listener);
    }

    /**
     * For now, just a on/off toggle, no real count used.
     */
    public void setBadgeCount(int count) {
        if (count == 0)
            ivBadge.setVisibility(View.GONE);
        else
            ivBadge.setVisibility(View.VISIBLE);
    }
}

File: plugin_src/browsertoolbar/src/ios/TitleBarIconView.swift
/*
* Copyright (c) 2020 Elastos Foundation
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/

import Foundation

public class TitleBarIconView: UIView {
    // UI
    @IBOutlet var iconView: AdvancedButton!
    @IBOutlet weak var badgeView: UIImageView!
    
    private var onClickListener: (()->Void)? = nil
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        let view = loadViewFromNib()
        self.translatesAutoresizingMaskIntoConstraints = false
        view.translatesAutoresizingMaskIntoConstraints = false
        addSubview(view)
        self.addMatchChildConstraints(child: view)
        
        setup()
    }

    required convenience init?(coder aDecoder: NSCoder) {
        self.init(frame: CGRect.null)
    }
    
    private func setup() {
        setBadgeCount(0)
    }
    
    /*public func setPaddingDp(int dpPadding) {
        int paddingPx = (int) Utility.convertDpToPx(getContext(), dpPadding);
        ivMainIcon.setPadding(paddingPx, paddingPx, paddingPx, paddingPx);
    }

    public func setImageResource(int resId) {
        ivMainIcon.setImageResource(resId);
    }

    public func setImageURI(Uri uri) {
        ivMainIcon.setImageURI(uri);
    }

    public func setColorFilter(int color) {
        ivMainIcon.setColorFilter(color);
    }

    public func setColorFilter(ColorMatrixColorFilter filter) {
        ivMainIcon.setColorFilter(filter);
    }*/

    public func setOnClickListener(_ listener: @escaping ()->Void) {
        onClickListener = listener
    }

    @IBAction func mainIconClicked(_ sender: Any) {
        onClickListener?()
    }
    
    /**
     * For now, just a on/off toggle, no real count used.
     */
    public func setBadgeCount(_ count: Int) {
        if (count == 0) {
            badgeView.isHidden = true
        }
        else {
            badgeView.isHidden = false
        }
    }
}
